Add ARM userspace infrastructure

agross-oss opened the pull request:
This patch set adds the ARM based userspace infrastructure.  The patches include all the parts to get system calls, demoting threads to user mode, etc.

andrewboie commented on arch/arm/core/thread.c:
```diff
@@ -67,22 +67,23 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
+#if CONFIG_ARM_USERSPACE
+	if (!(options & K_USER)) {
+		pInitCtx->pc = ((u32_t)_thread_entry) & 0xfffffffe;
```
why the mask on the function pointer?

agross-linaro replied:
I figured it out.  It's forcing the mode to ARM mode.  If the LSB is 1, it means it's thumb mode.

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+    mov lr,r0
+    push {r1,r2,r3,lr}
+
+    /* disable mpu stack guard to keep us from triggering guard region */
```
why is this needed?
On x86 we didn't need to touch the guard region at all.
Also please note, CONFIG_USERSPACE no longer depends on CONFIG_HW_STACK_PROTECTION and there may be no guard region.

agross-linaro replied:
yeah I just saw this.  I'll have to fix this piece.

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+    cmp r0,ip
+    blt valid_syscall
+    push {lr}
+    blx _do_kernel_oops
```
better to just override the system call id with K_SYSCALL_BAD and executing that entry in the dispatch table, with the bad ID as the first arg.
See x86 userspace.S

agross-linaro replied:
Yeah that is cleaner.  I'll change it to that.

andrewboie commented on arch/arm/core/fatal.c:
```diff
@@ -87,3 +87,9 @@ void _do_kernel_oops(const NANO_ESF *esf)
+
+FUNC_NORETURN void _arch_syscall_oops(void *ssf_ptr)
+{
+	_do_kernel_oops((const NANO_ESF *)ssf_ptr);
```
just wanted to double-check, is ssf_ptr really a NANO_ESF?
on x86 it wasn't and I had to copy stuff.

agross-linaro replied:
I'll double check.

agross-linaro replied:
i still have to fix this.  I'll push another change tomorrow to address the stack frame.  I think I'll have to copy the svc stack frame or reconstruct it from the pieces I have on hand (like the x86 method)

galak commented on arch/arm/core/thread.c:
```diff
@@ -67,22 +67,23 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 						     sizeof(struct __esf)));
 
+#if CONFIG_ARM_USERSPACE
+	if (!(options & K_USER)) {
```
can we swap this so its:

if (options & K_USER) {
   pInitCtx->pc = ((u32_t)_arch_user_mode_enter) & 0xfffffffe;
} else {
   pInitCtx->pc = ((u32_t)_thread_entry) & 0xfffffffe;
}

lpereira replied:
Maybe setting `pInitCtx->pc` to either `_arch_user_mode_enter` or `_thread_entry`, and masking the bits after the `#ifdef` block will generate slightly smaller code.

galak commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +327,19 @@ _oops:
     blx _do_kernel_oops
     pop {pc}
 
+_escalate_priv:
```
should we ifdef protect this?

agross-linaro replied:
the pop {pc} is nefarious because it literally pops off the stack into the pc, and then voila, you are there.  But yeah I can block this off.

tyrantlizard77 reviewed (changes requested):
I'm not too deeply familar with ARM internals but I do have some comments.

It's not clear to me how the kernel stack is set up, if this is being carved out of the existing thread stack, need to change this so that the kernel stack is reserved either completely separately (like on x86 where it is between the guard and the thread stack) or carved to a fixed size out of the thread stack, it shouldn't be a function of the thread stack size.

We don't know for sure how big the kernel stack needs to be, I'd suggest 2K as a reasonably safe choice and we can tune it later.

galak commented on include/arch/arm/arch.h:
```diff
@@ -307,54 +307,58 @@ typedef u32_t k_mem_partition_attr_t;
-	return 0;
+	u32_t control;
+	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(control));
+	return control & 0x1 ;
```
nit - remove space between 0x1 & ;

galak reviewed (dismissed):
some style comments

lpereira commented on include/arch/arm/arch.h:
```diff
@@ -307,54 +307,58 @@ typedef u32_t k_mem_partition_attr_t;
 static inline int _arch_is_user_context(void)
 {
-	return 0;
+	u32_t control;
```
Minor style nit in this function: missing vertical space between variable declaration and after the assembly block, and there's a stray space before the semicolon in the return statement.

andrewboie commented:
CI failures due to checkpatch being unhappy about whitespace

lpereira commented on include/arch/arm/arch.h:
```diff
@@ -307,54 +307,58 @@ typedef u32_t k_mem_partition_attr_t;
- * just for enabling CONFIG_USERSPACE on arm w/o errors.
+ * args land in the proper registers.
  */
+extern int _arm_do_syscall(u32_t call_id, u32_t arg1,
```
Will this work?  `call_id` is the last argument passed by all the `_arch_syscall_invoke*()` functions, and the assembly implementation of this routine seems to take `r0` as `call_id` (I'm not familiar with the ABI).

andrewboie replied:
good catch! yes this isn't correct and I think will fail since the call_id is in the wrong position

andrewboie replied:
this prototype seems to be no longer necessary, remove

andrewboie commented on include/arch/arm/arch.h:
```diff
@@ -307,54 +307,58 @@ typedef u32_t k_mem_partition_attr_t;
 static inline u32_t _arch_syscall_invoke0(u32_t call_id)
 {
-	return 0;
+        return _arm_do_syscall(0, 0, 0, 0, 0, 0, call_id);
```
One other thing worth considering: system calls are a very hot code path. Here we are spending cycles to marshal 0 values for unused parameters. Are there approaches where unused parameters can just leave undefined values in the registers instead?

lpereira replied:
If `call_id` is the first argument (that will always be set), the others don't need to be passed to the function.

agross-linaro replied:
i'll see if I can trim this down.

chunlinhan commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +327,19 @@ _oops:
+_escalate_priv:
+    /* check to see if we are already privileged */
+    mrs r2, CONTROL
+    tst r2, #0
```
Should be:
`tst r2, #1`
`beq _oops`

`tst r2, #0` the Z flag will always be 1.

Another question is, should we need oops for this situation?

andrewboie replied:
on x86 it is currently not a fatal error to invoke a system call in supervisor mode. it will work its just not a great way to go about using an API.

I don't think we should add any logic or checking to catch this. this is a *very* hot code path. it will not happen in practice unless the user is doing something really weird.

chunlinhan commented on arch/arm/core/thread.c:
```diff
@@ -93,5 +94,32 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+	_arm_userspace_enter(user_entry, p1, p2, p3,
+			     (u32_t)_current->stack_obj,
+			     _current->stack_info.size);
+	CODE_UNREACHABLE;
```
It would be better to add k_thread_abort() here like what _thread_entry() does to
handle the case that thread returns from user_entry() unexpectedly.

andrewboie replied:
NACK
that isn't how this works, user_entry will never return here

chunlinhan replied:
Yeah, I got it after reviewing the x86 implementation.

chunlinhan commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+    /* setup arguments to thread_entry function */
+    ldr r0,=_kernel
+    ldr r0,[r0, #_kernel_offset_to_current]
+    bl configure_mpu_stack_guard
```
I think the stack guard has been configured in __pendsv before _arch_user_mode_enter() being called.

andrewboie replied:
can we just make any adjustments needed to the guard area in C code and not in here?

i also don't have a clear picture on how the guard area, kernel stack, and thread stack are arranged

agross-linaro replied:
This gets complicated, because we have to use the guard region differently based on if we are in userspace or not (regardless of if STACK_GUARD is defined).  Basically from lowest to highest address the memory is laid out like:
|-----guard------|---------stack-----------------|
with userspace it'll be:
|------guard-----|-------------privileged stack------|-------unprivileged stack--------|

Added complexity is that the mpu regions have to be aligned.  So this means when we do userspace we absolutely have to have the stacks aligned to power of 2 boundaries.  This means the guard HAS to come out of the stack, not added to the stack.  All stacks have to fit in a power of 2 size.  The privileged stack also has to be power of 2.

About to push some fixes for this that changes the mpu regions to work properly for user space

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* return to thread_entry */
```
i don't think this is implemented correctly.

it looks like this instruction will directly enter the user_entry function.
This is incorrect. What we have to do is setup the stack/registers so that we enter _thread_entry() with user_entry and its three arguments as parameters to *that* function.
_thread_entry never returns, and calls k_thread_abort() and does other housekeeping if user_entry returns

agross-linaro replied:
bad comment.  This returns back to the point of the svc.  This does bring up the question of how we control who makes the call to svc.

chunlinhan replied:
I think what Andrew mentioned is that the better implementation should be like the following:

_arm_userspace_enter(k_thread_entry_t user_entry,
					       void *p1, void *p2, void *p3,
					       u32_t stack_end,
					       u32_t stack_start)
{
    .....
    .....
    _thread_entry(user_entry, p1, p2, p3);
}

but the current implementation is:

_arm_userspace_enter(k_thread_entry_t user_entry,
					       void *p1, void *p2, void *p3,
					       u32_t stack_end,
					       u32_t stack_start)
{
    .....
    .....
    user_entry(p1, p2, p3);
}

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+    push {r0,r1,r2,r3,r4,r5,r6,r7}
+    /* validate syscall limit */
+    ldr ip,=_SYSCALL_LIMIT
+    ldr r0, [sp, #40]
```
what does this magic value of 40 represent?
can we make this a macro instead?
also this looks very very wrong in that you are validating the system call ID **before** you elevate to privileged mode, which means the checks aren't useful from a security perspective.
you have to do *all* checking after the software interrupt is invoked, nothing done when the CPU is in user mode can be trusted

agross-linaro replied:
I can move the code to that.  That simplifies some things.

andrewboie commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +327,19 @@ _oops:
+    msr CONTROL, r2
+
+    /* fix PSP to point to privileged portion of stack */
+    bx lr
```
I'm a little confused about how this works..

It seems that upon "svc #3" we land here, and if we are unprivileged, we set mode to privileged, switch to the kernel stack, and return wherever execution was when the "svc #3" call was made.

what's to prevent malicious user code from entering privileged mode anytime they please and then running whatever code they want as a supervisor?

agross-linaro replied:
We can't do that unfortunately because we cannot handle nested SVC calls.  This would mean that we can't context switch without it faulting.

andrewboie commented:
Since we do not have emulator support, sanitycheck can't be used to verify this.

I suggest running the following test cases to validate the port. CONFIG_USERSPACE=y should be added to the defconfig for whatever platform you are testing on (ideally, one device with an ARM MPU and one with an NXP MPU)

```
tests/kernel/mem_protect/obj_validation/
tests/kernel/msgq/msgq_api/
tests/kernel/stack/stack_api/
tests/kernel/semaphore/sema_api/
tests/kernel/threads/lifecycle/thread_init/
tests/kernel/threads/custom_data/custom_data_api/
tests/kernel/mutex/mutex/
tests/kernel/alert/alert_api/
tests/kernel/threads/lifecycle/lifecycle_api/
```

Also test with CONFIG_HW_STACK_PROTECTION enabled and disabled.

I reviewed the code some more, I think I found a showstopper with how the privilege elevation is handled.

chunlinhan commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,154 @@
+#endif
+    mov r2,ip
+    sub r2,#24		/* don't erase the pushed regs */
+    bl memset
```
r0 will be stack_obj + stack_info.size here, so memset() will set the memory out of the range of stack_obj.

andrewboie replied:
has this been fixed?

agross-linaro replied:
yeah now we switch over to the priv stack and then memset out the complete user mode stack area.  No need to dance around any stored info here.

stephensmalley commented:
In case you don't already know it, if you try to build tests/kernel/msgq/msgq_api with CONFIG_USERSPACE=y and run it on frdm_k64f, it falls over with:

Program received signal SIGSEGV, Segmentation fault.
__usage_fault () at /home/sds/zephyr-project/arch/arm/core/fault_s.S:88
88		eors.n r0, r0
(gdb) where
#0  __usage_fault () at /home/sds/zephyr-project/arch/arm/core/fault_s.S:88
#1  <signal handler called>
#2  0x00001390 in _arch_syscall_invoke2 (call_id=57, arg2=22, arg1=536876204)
    at /home/sds/zephyr-project/include/arch/arm/arch.h:413
#3  k_str_out (n=22, 
    c=0x200014ac <_interrupt_stack+1932> "***** BUS FAULT *****\n")
    at /home/sds/zephyr-project/tests/kernel/msgq/msgq_api/build/frdm_k64f/zephyr/include/generated/syscalls/kernel.h:109
#4  buf_flush (ctx=0x200014a4 <_interrupt_stack+1924>)
    at /home/sds/zephyr-project/misc/printk.c:237
#5  0x00000000 in ?? ()

stephensmalley commented on arch/arm/core/swap.S:
```diff
@@ -283,12 +294,18 @@ SECTION_FUNC(TEXT, __svc)
+    cmp r1, #3
+    beq _do_syscall
+#endif
+
```
Would it be possible to check up front if the caller is unpriv and only allow it to use svc 3 in that case?  Otherwise, we're allowing userspace to invoke any of the other svc calls, and some of those might be trusting the caller.

agross-linaro replied:
Sure, the mode is in the CONTROL reg, and I did this check and bombed out in a previous version of this if the code was privileged.  I am on the fence on this one

stephensmalley replied:
My concern is different - I don't care about preventing a privileged caller from making an arbitrary svc call; I just want to prevent an unprivileged caller from calling anything other than svc 3 (and any other individual svcs that are explicitly whitelisted as being safe for use by userspace).

agross-linaro replied:
Ah ok I see what you mean.  This would mean some sort of macro or interface where we define system call numbers and their access mode and then do a lookup to validate it in the actual svc call.

andrewboie replied:
I don't think that's what Stephen is saying.
What we want to prevent is user mode from invoking (for example) SVC #1. That would be a backdoor into the system.

agross-linaro replied:
The whole point of the SVC is to run something as superuser.  That we use it today to do odd things like context switching or offloading while we are already privileged is kind or neither here nor there.  I don't know how you would enforce being able to service a SVC or not.  Sure we can add a lot of checks to make sure we are in the right 'state' when we make specific calls, but really,  I don't see what that gains us.  An SVC should do one specific thing and then return.

andrewboie replied:
Then I think we need to refactor the design.
Right now, we are using SVC for stuff which should never be allowable from user mode: irq_offload() and inducing a kernel panic. irq_offload runs an arbitrary function pointer in interrupt context. kernel panic hoses the entire system (unlike an oops which just kills a thread).
Is there no way to check what the interrupting context's privilege level was?

agross-linaro replied:
Yes you know the callers privilege level.  My point is that the SVC has always been meant to be called from unprivileged contexts.  If we want to add this extra consideration, so be it.

andrewboie replied:
> My point is that the SVC has always been meant to be called from unprivileged contexts

For the irq_offload() and panic cases, we are using SVC because we want to generate a CPU exception from software, even though these should not be allowable from user mode.
Is there perhaps a better way to do it?
For sake of efficiency, we could check that the CPU is not in user mode in the logic for panic/offload cases and jump to the oops case if the test fails.

agross-linaro replied:
And what do we return if we invoke a syscall from privileged context?  We shouldn't allow that either.

andrewboie replied:
> And what do we return if we invoke a syscall from privileged context? We shouldn't allow that either.

Do we care?
AFAIK on x86 it does work, although it's a terribly inefficient thing to do since you can just call the implementation directly.
I'm fine with leaving the scenario as "undefined behavior" and not doing checks, or only checking in an assertion. I can't conceive on how supervisor mode would be invoking system calls like this unless someone is doing something really weird, and we don't try to prevent malfeasance in supervisor mode anyway.

stephensmalley commented:
What test(s) are you running to exercise this? As I mentioned above, msgq_api test falls over with a usage fault on frdm_k64f with CONFIG_USERSPACE=y.
Also, latest commits seem to introduce a build error for several tests on frdm_k64f, e.g. for stack_api with CONFIG_USERSPACE=y,
```
In file included from /home/sds/zephyr-project/include/arch/cpu.h:15:0,
                 from /home/sds/zephyr-project/include/kernel.h:33,
                 from /home/sds/zephyr-project/include/zephyr.h:18,
                 from /home/sds/zephyr-project/kernel/init.c:14:
/home/sds/zephyr-project/include/arch/arm/arch.h:139:9: error: requested alignment is not a positive power of 2
  struct _k_thread_stack_element __kernel_noinit __aligned(size) \
         ^
/home/sds/zephyr-project/include/kernel.h:4092:42: note: in expansion of macro ‘_ARCH_THREAD_STACK_DEFINE’
 #define K_THREAD_STACK_DEFINE(sym, size) _ARCH_THREAD_STACK_DEFINE(sym, size)
                                          ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/sds/zephyr-project/kernel/init.c:85:1: note: in expansion of macro ‘K_THREAD_STACK_DEFINE’
 K_THREAD_STACK_DEFINE(_idle_stack, IDLE_STACK_SIZE);
 ^~~~~~~~~~~~~~~~~~~~~
```

agross-linaro commented:
Sorry I should have posted that I was pushing this to give access to another engineer for changes.  I'll do that in the future.  I'll also post known issues.

The usage_fault in the msg q api is actually a bug in the msq q api.  If you trace back the usage_fault, you'll see it's a segmentation fault that shows that a user mode thread is trying to read from a location on its stack which resides outside the user mode thread's stack region and accessible memory partitions.  This is not allowed without adding the structure being accessed into a memory domain that the user mode thread can access.  I don't understand why this is successfully running on x86.  I'll have to dig further.

agross-linaro commented:
Sorry I should have posted that I was pushing this to give access to another engineer for changes.  I'll do that in the future.  I'll also post known issues.

The usage_fault in the msg q api is actually a bug in the msq q api.  If you trace back the usage_fault, you'll see it's a segmentation fault that shows that a user mode thread is trying to read from a location on its stack which resides outside the user mode thread's stack region and accessible memory partitions.  This is not allowed without adding the structure being accessed into a memory domain that the user mode thread can access.  I don't understand why this is successfully running on x86.  I'll have to dig further.

chunlinhan commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +341,41 @@ _oops:
+    bge skip_raising_privilege
+
+    /* set mode to privileged */
+    bic r2, #1
```
should add
mrs r2, CONTROL
before clearing bit 0 of r2

chunlinhan commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +341,41 @@ _oops:
+    ldr ip,=_SYSCALL_LIMIT
+    ldr r1, [r0, #40]
+    cmp r1,ip
+    bge skip_raising_privilege
```
I think we still need to raise privilege even the id is over _SYSCALL_LIMIT.
Because we still need the privilege to execute _handler_bad_syscall() later.

agross-linaro replied:
I've tested the syscall bad path and it appears to execute fine in user mode and prints out the error messages just fine.

chunlinhan replied:
yeah, I forget that when we do printk in _handler_bad_syscall(), the privilege will be raised if we are in user mode.
It seems it won't hurt if we always do raise prilivege here and it can also reduce conditional branch.

chunlinhan commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,178 @@
+    str r6, [r5, #12]	/* store ssf of unprivileged stack */
+
+    /* switch to privileged stack */
+    msr PSP, r5
```
I do not see the needs that we need line 118 ~114, since we are already in thread mode here and all things
is on PSP now. I've tried to remove related code here and in __pendsv and tested k_sleep() and printk(), everyting
seems works fine too.

stephensmalley commented:
> The usage_fault in the msg q api is actually a bug in the msq q api test. If you trace back the usage_fault, you'll see it's a segmentation fault that shows that a user mode thread is trying to read from a location on its stack which resides outside the user mode thread's stack region and accessible memory partitions. This is not allowed without adding the structure being accessed into a memory domain that the user mode thread can access. I don't understand why this is successfully running on x86. I'll have to dig further.

Hmm...I thought it was failing on the attempt to access the test_msgq_api unit_test structure, which is static and not on the stack at all. Is this related at all to the fact that x86 enables CONFIG_APPLICATION_MEMORY and configures its MMU to allow user threads access to application globals? Not clear that's implemented for ARM presently (linker script support is there, but is the MPU configured to allow it?).

agross-linaro commented:
Yeah Chunlin brought up that point last night.  The x86 sets APPLICATION_MEMORY by default these days, which shunts all of the offending variables into the apps bss/data.  They must be setting those pages to user accessible.  That's a missing piece that needs to be filled in for ARM.

We're in the process of fixing that bit, so the next patch set will include that.  The implementation of that will involve a couple of things:
1) Modifying the linker scripts to align the app data/bss section to something that is MPU friendly.  Some MPUs require alignment to power of 2 boundaries.  That requires determining the size of the section, aligning it to the next power of 2, and padding off the remainder to make sure the linker doesn't place anything else there.
2) Adding the app data/bass partition to the context switch mechanism to make sure it's accessible.

stephensmalley commented:
FWIW, I have started on a set of tests for the security guarantees allegedly provided by CONFIG_USERSPACE=y.  WIP can be seen in https://github.com/zephyrproject-rtos/zephyr/pull/5111. Can't yet exercise these patches until the CONFIG_APPLICATION_MEMORY support is there and build issues are sorted.

agross-linaro commented:
That's quite nice.  I'll go through the wip later this evening after some holiday travel.

chunlinhan commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,178 @@
+    msr PSP, ip
+
+    /* reset stack permissions to allow for unprivileged access */
+    ldr r0, =_kernel
```
Need to keep r0 (return value of the syscall) before we destroy it.

agross-linaro replied:
good catch.

andrewboie commented on arch/Kconfig:
```diff
@@ -69,6 +69,16 @@ config USERSPACE
+	depends on ARCH_HAS_USERSPACE
+	help
+	This option sets the priviliged stack region size that will be cut
+	out of the user mode thread stack.  During normal execution, this
```
Would it be possible to *add* this space to the thread stack, instead of cutting space out?
Otherwise, turning on user mode will have the effect of subtracting 512 bytes of stack space from every thread.
We want the app to work more or less the same if usermode is or is not enabled.

agross-linaro replied:
It breaks power of 2 alignment, so no.  This would cause us to have to double the size of stacks and then waste a bunch of space on the end, which btw would be accessible by the thread.  The only way around that is to then define additional guards, which expends regions we can't afford to expend.

andrewboie replied:
We need to express this in Kconfig in some different way then. We need to make this really clear to users what is going on.

Not all MPUs enforce regions to be aligned to a multiple of their power-of-two size, and MMU-based systems just add an extra page. Is there some way we can set this up so that this 'carving out' behavior only takes place on systems with very strong region alignment constraints?

Maybe another Kconfig which controls if the kernel stack is added or carved out?

agross-linaro replied:
this is more a function of the mpu features.  I think we need at least a Kconfig option for the alignment characteristic of the mpu.  Vanilla ARM mpus (ST and Beetle) need power of 2.  The NXP is the only one that only requires 32 byte alignment.  Mainly due to the fact they use a start/end pair of registers for defining the region.

andrewboie replied:
Adding another Kconfig for these characteristics is fine w/me.
Once this all lands, I can update the x86 port to use this value (and default to 4096) instead of hard-coding the kernel stack size, for sake of consistency.

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,178 @@
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* jump to user entry */
```
I might be mis-reading this, but it appears we are directly diving into the supplied user_entry function.
We shouldn't be doing that.
We should be "returning" to the kernel function _thread_entry() define in kernel/thread.c with user_entry, arg0, arg1, arg2 as its arguments.

agross-linaro replied:
How would we return to "_thread_entry()"?  We never invoked that function.  We set the entry to user_mode_enter, which calls the asm function which in turn calls the user supplied function.  The arch_user_mode_enter() is basically a copy of the _thread_entry, except it calls the asm function to drop to user mode and jump to the supplied function.  Are you implying that we need to call _thread_entry after dropping to user mode?  Which would mean returning back to the arch user mode enter, and then calling _thread_entry?

andrewboie commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,178 @@
+     */
+    push {r4,r5,r6,lr}
+
+    ldr r4, =_SYSCALL_LIMIT
```
we already validated the system call ID in _do_syscall, why are we doing it again here?

agross-linaro replied:
because at the end of the svc, all of the registers are reset back to what they were before and the exception frame is popped off the stack.  So at the point where we enter the actual syscall, we have r0/r1/r2/r3 holding arguments, and the stack holding arg5/arg6/call_id/lr.  If I overwrite the call_id with the SYSCALL_BAD, then we lose the call_id information.

I suppose I could go ahead and overwrite the arg0 with the bogus sycall value, and overwrite the call_id stack entry with the SYSCALL_BAD and just dispatch it like usual.

Commit:
arm: mpu: Add user context API placeholder

This patch adds a configure_mpu_user_context API and implements
the required function placeholders in the NXP and ARM MPU files.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
arm: userspace: Add ARM userspace infrastructure

This patch adds support for userspace on ARM architectures.  Arch
specific calls for transitioning threads to user mode, system calls,
and associated handlers.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
arm: mpu: Enable userspace support for NXP and ARM

This patch set implements the APIs and changed required to support
the user mode thread support.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
tests: mem_protect: userspace: Adjust kernel stack tests

This patch adjusts the calculation of the overflow size for the kernel
stack tests which read/write to areas below the current user stack.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

agross-linaro commented on scripts/gen_alignment_script.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2017 Intel Corporation
```
oops wrong attribution.  ill fix this.

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -31,18 +32,66 @@ static inline u32_t _get_region_attr(u32_t xn, u32_t ap, u32_t tex,
 		| (c << 17) | (b << 16) | (srd << 5) | (size));
 }
 
+static inline u32_t round_up_to_next_power_of_two(u32_t v)
```
actually can we just use POW2_CEIL() defined later in this PR since it does the same thing?

lpereira replied:
Indeed. This trick to round up is a well known bit twiddling hack, but the code using `__builtin_clz()` is often tighter.

agross-linaro replied:
yeah we should be able to sub in the pow2_ceil usage here.

andrewboie replied:
i see you have a fixup patch for this later in the series, I'd like to see that squashed into this one.

andrewboie commented on include/arch/arm/arch.h:
```diff
@@ -306,55 +354,171 @@ typedef u32_t k_mem_partition_attr_t;
 					  u32_t arg4, u32_t arg5, u32_t call_id)
 {
-	return 0;
+	register u32_t ret __asm__("r0") = arg1;
```
are these actually necessary?
does it do the same thing if you just put the arguments directly in the __asm__ call? the "r" constraint should ensure that they end up in registers

andrewboie replied:
disregard, we need these to end up in *specific* registers and ARM doesn't have constraint codes for them

tyrantlizard77 reviewed (changes requested):
A few comments...not finished reviewing yet. I need to look over the new stack handling stuff in detail.

tyrantlizard77 reviewed (commented):
I think I understand how the privileged stack generation works now.
I'd like to see doc/kernel/usermode/kernelobjects.rst updated with a high-level overview of what is going on.
Also, if we enable HW stack protection, this generates stack guard regions for kernel stacks, which are either the stack itself for non-user threads, or the privileged stack for threads that are doing a system call. How is the guard region allocated in terms of the specified stack size? Is it carved out of the beginning? That should be OK since the guards are small, but we may want to document it.

agross-linaro commented on arch/Kconfig:
```diff
@@ -69,6 +69,16 @@ config USERSPACE
 
+config PRIVILEGED_STACK_SIZE
+	int "Size of privileged stack"
+	default 256
```
It's a guess.  We can change the defaults if we run into issues.  Aside from network calls, i have to wonder how deep some of these call chains are.

andrewboie replied:
wrt network stack I think we can cross that bridge when we get there, as there is currently no network functionality exposed via system calls.
if 256 is a wild guess, let's make sure we have stack guards turned on when we test, so that we know when we have overflowed.

agross-linaro commented:
Will look into creating the python lib for this stuff.

stephensmalley commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +352,47 @@ _oops:
+    ldr r1, =_arm_do_syscall
+    str r1, [r0, #24]   /* overwrite the LR to point to _arm_do_syscall */
+
+    /* validate syscall limit, only set priv mode if valid */
```
Second part of this comment is no longer true.

stephensmalley replied:
So why bother validating here at all, and not just do it in _arm_do_syscall?

agross-linaro replied:
I'll fix the comment.  So in a previous review I was doing just that, but there was a comment about making sure not to progress any further into the SVC without validating.  I was having to do the syscall check in both places because I am limited to what I can store or overwrite because of the nature of the SVC call/return.  So I opted to do it here.

stephensmalley replied:
If we are going to do the priv esc regardless, I'm not sure what we are gaining by doing the check here vs in _arm_do_syscall itself (it would make more sense if we were going to oops or otherwise fail immediately w/o privilege elevation and w/o calling _arm_do_syscall at all).  The risk of only checking here is that an inconsistency could arise between this check and the actual use by _arm_do_syscall. Also not sure you want _arm_do_syscall to be a gadget for calling any function pointer in memory after the system call table, since it is fully trusting its caller to have done the validation and set SYSCALL_BAD a priori.

agross-linaro replied:
Calling any function pointer in memory would only be the case if the thread making the call to the _arm_do_syscall is privileged.  If the thread was unprivileged, it'd fault at the point of dealing with the privileged stack as it would not have the permissions to access that memory region.  And this would occur before the branch to the function ptr.

stephensmalley commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +352,47 @@ _oops:
+    msr CONTROL, r2
+
+    /* return from SVC to the modified LR - _arm_do_syscall */
+skip_raising_privilege:
```
This label is no longer used.

agross-linaro replied:
thought i had removed that.  I'll fix that.  thanks

stephensmalley commented on arch/arm/core/swap.S:
```diff
@@ -324,6 +352,47 @@ _oops:
+    ldr r1, [r0, #24]   /* grab address of PC from stack frame */
+    str r1, [r0, #44]   /* store address to use for LR after syscall */
+    ldr r1, =_arm_do_syscall
+    str r1, [r0, #24]   /* overwrite the LR to point to _arm_do_syscall */
```
So, nothing else can overwrite this in order to redirect control flow after privilege elevation?  We're operating on the unprivileged stack but relying on the fact that no other thread can access it?  Why do we refer to [r0, #24] as PC above but LR here?

agross-linaro replied:
I should fix this comment.  What we are doing is storing off the PC that would have been used in the return into a LR storage spot we set aside.  Then we are overwriting the PC to return from interrupt into the _arm_do_syscall.

stephensmalley replied:
So, to be clear, the security of this depends on the fact that no other thread can write to this thread's (user) stack and there is no way that the overwritten PC on the stack frame could be clobbered with a different value prior to the bx lr?

agross-linaro replied:
At the point that this is running, we are in exception mode doing a SVC.  Only a higher priority interrupt could get inbetween us making this modification and the bx lr which pops all this information off the stack and returns to the new PC.

lpereira replied:
SMP Zephyr is going to be a reality soon. Is there any multicore ARM SoC where this would be an issue?

stephensmalley commented on arch/arm/core/swap.S:
```diff
@@ -176,12 +177,24 @@ _thread_irq_disabled:
+
+    add r0, r2, #0
+    push {r2, lr}
+    blx configure_mpu_user_context
```
You didn't start this trend, but wondering if all of these separate calls to configure_mpu_*, each with its own disable/configure/enable sequence, is a good idea.  Performance overhead of disabling/enabling the MPU for each such region?

agross-linaro replied:
performance overhead is a single register write to enable/disable the regions globally.  This would be equal to or less costly than toggling 1 or more region enables.

agross-linaro replied:
I thought about collapsing all of the mpu modifications into a single function, but then decided against it mainly because of the separate nature of the features.

stephensmalley commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,167 @@
+    mov r6, sp
+    str r6, [r5, #8]	/* store ssf of unprivileged stack */
+    ldr r6, =0
+    str r6, [r5, #12]	/* store ssf of unprivileged stack */
```
cut-and-paste on the comment (doesn't match the code).

agross-linaro replied:
will fix

stephensmalley commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,167 @@
+
+    msr PSP, r4
+
+    /* setup arguments to thread_entry function */
```
Comment doesn't match code

agross-linaro replied:
will fix

stephensmalley commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,167 @@
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl configure_mpu_mem_domain
+
+    /* setup arguments to thread_entry function */
```
Ditto

agross-linaro replied:
will fix

stephensmalley commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -144,8 +144,12 @@ static void write_kernel_data(void)
+#else
+volatile size_t size = CONFIG_ZTEST_STACKSIZE + MPU_GUARD_ALIGN_AND_SIZE;
+#endif
 #else
```
The actual intent of the read/write_kernel_stack tests is to try to write to the kernel/privileged stack for the user thread, not just to write below the current user stack.  The logic here assumed that the privileged stacks were carved out of the user stacks.  If we have to instead calculate their location independently, then these two tests need to be reworked accordingly.

agross-linaro replied:
There is no way to calculate this now because there is no way to determine this unless we want to expose APIs to look up the privileged stack area from unprivileged threads.

stephensmalley replied:
It appears that if I add -C to the gperf command line for generating the priv stacks hash table (so that it ends up in rodata), I can call _k_priv_stack_find(thread_stack) from application code and get back the priv stack pointer, and then attempting to dereference that pointer faults as expected. (That also required making thread_stack in ztest.c non-static so I could reference it from the test.) OTOH, not clear that we want to expose the priv stack locations to userspace if we can avoid it.

stephensmalley replied:
FWIW, this can be seen at https://github.com/stephensmalley/zephyr/commit/622028c5e8c3ec1db61379bf67a05d657d510775 relative to your patches.  As I said though, not sure if we want to do this.

andrewboie replied:
> not clear that we want to expose the priv stack locations to userspace if we can avoid it.

I'll let you guys decide this, my feeling is that a determined hacker is gonna figure this out anyway so we should assume it's always known.

stephensmalley replied:
Re-based fix for the read/write_stack tests at https://github.com/stephensmalley/zephyr/commit/2d2273cfd25afee096e5ff557e674a7b23d2df6a.  Feel free to add to this PR, either on top or squashed.

stephensmalley replied:
Ok, how about this one then: https://github.com/stephensmalley/zephyr/commit/1933d4e21fff5232d43cb916e0003f6b5d70968f

stephensmalley commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -144,8 +144,12 @@ static void write_kernel_data(void)
+#endif
 #else
 volatile size_t size = 512;
 #endif
```
Also, let's just make this #else clause an #error "Not implemented for this architecture".

andrewboie replied:
agreed

chunlinhan commented on arch/arm/core/thread.c:
```diff
@@ -67,31 +78,68 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
+#if CONFIG_ARM_USERSPACE
+	thread->arch.mode = 0;
+	if (options & K_USER) {
```
Can't set priv_stack_start and priv_stack_size as zero here here even it is a privileged thread, since a privileged thread could drop it privilege later by calling k_thread_user_mode_enter().

chunlinhan replied:
Maybe we could levave the code here as it is but just add extra code in _arch_user_mode_enter() to
set priv_stack_start and priv_stack_size when a privileged thread wants to drop its privilege.

agross-linaro replied:
that's fine.  is that even a use case though running as privileged for a while then dropping?  I thought this was all decided at the point the thread is created

chunlinhan replied:
I think sometimes we may create threads as privileged and drop their privilege after setting up memory domain
or access right of kobjects.

andrewboie replied:
> e. is that even a use case though running as privileged for a while then dropping

Yes it definitely is.
The API is k_thread_user_mode_enter()

agross-linaro commented:
Rebased.  Two things on this set of patches:
1) There is an issue with computing the end of a stack when the stack has a guard, userspace is enabled, and the mpu requires power of two alignment.  I am going to add in a patch to add a K_THREAD_STACK_END macro for users to use instead of computing their own using the K_THREAD_STACK_BUFFER and size of their stack.

2) I need to convert over the priv_stacks script to use the elf helper.

Both of these will be completed in the next pull request

andrewboie commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -194,64 +243,42 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
+	size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start;
+
+	/* set up app data region if exists, otherwise disable */
+	if (size > 0)
```
zephyr coding style insists on braces for all if clauses even if they are one-liners

tyrantlizard77 reviewed (changes requested):
Looking at this latest PR:

1) Seeing some patches in here which are fixups with no Signed-off-by. Need to squash these into whatever commits are necessary, or rewrite the commit messages.

2) Some comments from older iterations of this PR don't seem to be addressed yet, seems to be mostly minor stuff about missing/incorrect comments and whatnot, I'd go through them all again.

3) Really need a commit to doc/ with some human-readable detail implementation details on how the whole privileged stacks / power of 2 aligned stacks work. Don't need to dive super deep into it but a couple or three paragraphs would be really helpful to understand what is going on here.

Commit:
arch: arm: Fixup stack end calculations

This patch fixes calculations for the top of the interrupt and main
stacks.  Due to power of two alignment requirements for certain MPUs,
the guard size must be taken into account due to the guard being
counted against the initial stack size.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
doc: kernel: usermode: Add MPU stack and userspace documentation

This patch adds documentation on the design and implementation of stack
objects for architectures which utilize MPU backed stack and memory
protection.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

andrewboie commented:
Reviewing this now, but there are a ton of CI failures that need to be looked into.

agross-linaro commented:
yeah i am looking at the sanitycheck to get valgrind info on why this is failing.

tyrantlizard77 reviewed (changes requested):
Some minor stuff, I think we are getting really close.

About K_THREAD_STACK_SIZEOF and related macros, I have spent some time thinking about this and I don't think we need to worry about doing anything drastic or change the API.

K_THREAD_STACK_SIZEOF's only constraint is that it returns the same value passed into K_THREAD_STACK_DEFINE, and it is intended only to pass a size to k_thread_create(), it is not supported for any other use. strictly speaking, if userspace is enabled we don't even need a size argument to k_thread_create since we can snarf the stack size out of the stack's associated struct _k_object.data field, but this is to be consistent with non user mode configurations. It's already documented that it doesn't work with K_THREAD_STACK_ARRAY_DEFINE.

There are no specific guarantees that any guard area is added to or carved out of the size passed to K_THREAD_STACK_DEFINE. On x86 it is added, but on ARM the usable stack area will have 32 bytes subtracted from it. I'm fine with this policy of the kernel potentially reserving some of the provided size for internal purposes.

The only thing that we need to make sure happens is that inside the _new_thread() implementation, the initial pointer and size of the *actual* usable area is passed to _new_thread_init(). These values get stashed in the thread->stack_info and any runtime logic to analyze the true stack buffer will use this data.

also, K_THREAD_STACK_BUFFER should return the starting pointer of the usable area, so just add the guard size and cast to char *.

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -240,6 +246,31 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
+	/* configure stack */
+	_region_init(index, base, region_attr);
+
+	/* configure app data portion */
```
This might need to be ifdef'd around CONFIG_APPLICATION_DATA

agross-linaro replied:
done

stephensmalley replied:
It's CONFIG_APPLICATION_MEMORY last I looked, not _DATA.

lpereira commented on arch/arm/core/thread.c:
```diff
@@ -63,35 +67,71 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
+#if CONFIG_ARM_USERSPACE
+	thread->arch.mode = 0;
+	thread->arch.priv_stack_start = 0;
```
Is it worth the trouble of branching instead of zeroing out two ints in a struct?

agross-linaro replied:
probably not worth the trouble, imho

lpereira commented on arch/arm/core/userspace.S:
```diff
@@ -0,0 +1,177 @@
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0x0
```
Would "EOR r1, r1, r1" be more compact?

agross-linaro replied:
yeah, i'll change it to eors.n r1, r1

lpereira commented on scripts/elf_helper.py:
```diff
@@ -0,0 +1,522 @@
+# --- type classes ----
+
+
+class KobjectInstance:
```
gen_kobject_list.py declares these classes as well; can we move those to a separate file and only import them in this script? (Same for KobjectType, ArrayType, etc.)

andrewboie replied:
agreed

lpereira replied:
Oh, disregard this comment. This is precisely what this file is.

agross-linaro replied:
yeah i split the change into 2 separate patches.  1 to implement the helper, and the other to change usage of existing files over to use the new helper.

andrewboie commented:
@dbkinder some doc changes in this latest iteration.

andrewboie commented on CMakeLists.txt:
```diff
@@ -636,6 +635,119 @@ if(CONFIG_GEN_ISR_TABLES)
+  add_custom_command(
+    OUTPUT ${PRIV_STACKS_OUTPUT_SRC_PRE}
+    COMMAND
+    ${GPERF}
```
add -C so we can put this stuff completely in ROM

agross-linaro replied:
done

andrewboie commented on scripts/gen_priv_stacks.py:
```diff
@@ -0,0 +1,123 @@
+footer = """%%
+u8_t *_k_priv_stack_find(void *obj)
+{
+    struct _k_priv_stack_map *map = _k_priv_stack_map_lookup((const char *)obj,
```
can be const

agross-linaro replied:
done

agross-linaro commented on arch/arm/core/fatal.c:
```diff
@@ -87,3 +87,14 @@ void _do_kernel_oops(const NANO_ESF *esf)
+	u32_t *ssf_contents = ssf_ptr;
+	NANO_ESF oops_esf = { 0 };
+
+	oops_esf.pc = ssf_contents[3];
```
No it's not stashed by hw.  We'd have to implement that at the point we do the svc.

agross-linaro commented on scripts/gen_alignment_script.py:
```diff
@@ -0,0 +1,74 @@
+def parse_args():
+    global args
+
+
```
done

agross-linaro commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -182,6 +184,10 @@ static inline int _is_user_accessible_region(u32_t r_index, int write)
 	r_ap = ARM_MPU_DEV->rasr & ACCESS_PERMS_MASK;
 
+	/* always return true if this is the thread stack region */
+	if (_get_region_index_by_type(THREAD_STACK_REGION) == r_index)
```
i missed one.  fixed it.  thanks

dbkinder commented on doc/kernel/usermode/arm_userspace.rst:
```diff
@@ -0,0 +1,26 @@
+guard (if applicable).
+
+During system calls, the user mode thread's access to the system call and the
+passed in parameters are all validated.  The user mode thread is then elevated
```
``passed-in`` (with a hyphen)

agross-linaro replied:
fixed

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+#################
+
+Thread Stack Creation
+=====================
```
our convention is to use asterisks for H2 headings  (equal signs are for H3 headings)

agross-linaro replied:
ill change it to *

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+implementation of other features such as stack protection and userspace.
+
+Stack Guards
+============
```
asterisks here  (****)

agross-linaro replied:
done

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+or the lowest address.
+
+Memory Placement
+================
```
asterisks here (****)

agross-linaro replied:
done

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+
+For architectures which utilize memory protection unit (MPU) hardware,
+stacks are physically contiguous allocations.  This contiguous allocation
+has implications for the placement of stacks in memory, as well as the
```
what are the implications?

agross-linaro replied:
added a sentence pointing to the memory placement section later in the document.

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+to memory.  Memory protection units can provide this kind of support.
+The MPU provides a fixed number of regions.  Each region contains information
+about the start, end, size, and access attributes to be enforced on that
+particuliar region.
```
``particular``

agross-linaro replied:
fixed

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+Stack guards are implemented by using a single MPU region and setting the
+attributes for that region to not allow write access.  If invalid accesses
+occur, a fault ensues.  The stack guard is defined at the bottom of the stack,
+or the lowest address.
```
To be clear, I think you're saying,

    The stack guard is defined at the bottom (the lowest address) of the stack.

You're use of "or" makes it sound like an alternative rather than clarifying what the "bottom of the stack" means.

agross-linaro replied:
right.  i'll change it to your suggestion

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+The main source of the memory constraints is the MPU design for the SoC.
+Some MPUs require that each region be aligned to a power of two.  These SoCs
+will have :option:`CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT` defined.
+This means that a 1500 byte stack should be aligned to a 2kB boundary.  This
```
I'd rewrite this sentence a bit, and move it to the end of this paragraph:

    This requires a power of two ceiling be computed for the alignment and size
    of the stack, and can result is the actual memory allocation of a given stack
    being larger than the requested size.  For example, a request for a 1500 byte stack
    would be aligned and resized to a 2kB boundary and size if a power of two alignment
    is required by the MPU.

agross-linaro replied:
i redid the section.   Take a look at the new version.

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+stack.  The result is that the actual memory allocation of a given stack may
+be larger than the request.
+
+For MPUs which do not have a power of two alignment constraint, the minimum
```
s/which/that/

agross-linaro replied:
fixed

dbkinder commented on doc/kernel/usermode/mpu_stack_objects.rst:
```diff
@@ -0,0 +1,49 @@
+be larger than the request.
+
+For MPUs which do not have a power of two alignment constraint, the minimum
+alignment is required to be 32 bytes.
```
is there a size implication here too, such as multiples of 32 bytes?

agross-linaro replied:
No.  those implementations use a start/end address register.  The only requirement is that start/end are aligned on 32 bytes.

agross-linaro replied:
This sentence probably needs to be a little more precise.

dbkinder reviewed (dismissed):
some questions and suggested edits

stephensmalley commented:
This appears to be "passing" the userspace tests but for the wrong reason.
Faulting during run_test_functions() upon the attempted dereference of the unit_test, probably an issue with your CONFIG_APPLICATION_MEMORY support.  So we fault on every test but before we reach the actual test code.  I'm going to add explicit setting and checking of an expect_fault boolean around all of the tests so we don't encounter this again.

agross-linaro commented:
So the userspace tests need APPLICATION MEMORY set to make sure they can access the shared user data section.  Otherwise all they get access to is the thread stack and defined memory domains.

stephensmalley commented:
Yes, userspace has CONFIG_APPLICATION_MEMORY=y in its prj.conf.  That didn't change.

andrewboie commented:
>  So the userspace tests need APPLICATION MEMORY set to make  sure they can access the shared user data section.  Otherwise all they  get access to is the thread stack and defined memory domains.

Yes.
It's turned on in qemu_x86_defconfig.
It might be better to instead explicitly state this in the prj.conf for all the tests which can work with userspace enabled.

andrewboie commented:
> This appears to be "passing" the userspace tests but for the wrong reason.

Ouch. That sounds like a bad bug in ztest.

stephensmalley commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -194,64 +244,46 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
+	/* configure stack */
+	_region_init(index, base, ENDADDR_ROUND(base + size), region_attr);
+
+#if defined(CONFIG_APPLICATION_DATA)
```
Also here

agross-linaro replied:
Ack.  Will fix.

andrewboie commented:
This needs a rebase.

andrewboie commented:
There was at one point a patch against doc/ which described how the privileged stack allocation works. But I can't seem to find it any more. Can you re-add it?

andrewboie commented on tests/ztest/src/ztest.c:
```diff
@@ -198,7 +198,7 @@ static int run_test(struct unit_test *test)
 
 	TC_START(test->name);
 	k_thread_create(&ztest_thread, thread_stack,
-			K_THREAD_STACK_SIZEOF(thread_stack),
```
Something is really wrong here.
thread_stack is defined as: 

```
K_THREAD_STACK_DEFINE(thread_stack, CONFIG_ZTEST_STACKSIZE +
 			     CONFIG_TEST_EXTRA_STACKSIZE);
```

K_THREAD_STACK_SIZEOF(thread_stack), by definition, returns the same value passed to K_THREAD_STACK_DEFINE. It should already be  CONFIG_ZTEST_STACKSIZE + CONFIG_TEST_EXTRA_STACKSIZE.
Why are you getting a different result?

andrewboie commented on doc/kernel/usermode/arm_userspace.rst:
```diff
@@ -0,0 +1,26 @@
+.. _arm_userspace:
+
+ARM Userspace
```
Still need to change this to not be ARM-specific. ARC port is using this too, for example.

tyrantlizard77 reviewed (changes requested):
Just a few issues.
This is gonna need a bunch of testing but I'm inclined to merge this soon.

agross-linaro commented:
addressed Andrew's issues with the ztest change and the docs being ARM specific.  The ztest change ended up being unnecessary due to a later addition in the arm specific _new_thread changes.

tyrantlizard77 reviewed (approved):
Gonna push this once we have a positive CI result.
